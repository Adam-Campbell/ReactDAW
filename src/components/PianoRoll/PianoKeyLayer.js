import React from 'react';
import { Layer, Rect } from 'react-konva';
import PropTypes from 'prop-types';
import { UIColors } from '../../constants';

/**
 * Handles the click of a piano key.
 * @param {Object} e - the event object
 * @param {String} pitch - the pitch of the key that was clicked
 * @param {String} channelId - the id of the channel this section belongs to
 */
const handlePianoKeyClick = (e, pitch, channelId) => {
    // Trigger a note to play on the instrument corresponding to this channel, but do it by directly 
    // interacting with the instrument rather than going through Redux, since this is not state that
    // should be persisted in any way. 
    e.cancelBubble = true;
    window.instrumentReferences[channelId].triggerAttackRelease(pitch, '8n');
};

const PianoKeyLayer = props => (
    <Layer
        y={40}
        ref={props.pianoKeyLayerRef}
    >
        {props.pitchesArray.map((pitch, index) => (
            <Rect 
                x={0}
                y={index * 16}
                width={48}
                height={16}
                stroke={UIColors.backgroundDark}
                strokeWidth={2}
                fill={pitch.includes('#') ? UIColors.backgroundMedium : UIColors.offWhite} 
                key={index} 
                pitch={pitch}
                type={'pianoKeyRect'}
                onClick={e => handlePianoKeyClick(e, pitch, props.channelId)}
                onMouseDown={e => { e.cancelBubble = true; }}
                onMouseUp={e => { e.cancelBubble = true; }}
            />
        ))}
    </Layer>
);

PianoKeyLayer.propTypes = {
    pianoKeyLayerRef: PropTypes.object.isRequired,
    pitchesArray: PropTypes.arrayOf(PropTypes.string).isRequired,
    channelId: PropTypes.string.isRequired
};

export default PianoKeyLayer;