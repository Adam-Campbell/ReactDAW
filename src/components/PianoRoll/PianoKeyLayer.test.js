import React from 'react';
import PianoKeyLayer from './PianoKeyLayer';
import { Rect } from 'react-konva';
import { shallow } from 'enzyme';
import { createPitchesArray } from '../../sharedUtils';

const mockedHandleClick = jest.fn();

test('renders correctly', () => {
    const component = shallow(
        <PianoKeyLayer 
            pianoKeyLayerRef={React.createRef()}
            pitchesArray={createPitchesArray()}
            channelId="9132467532619432"
        />
    );
    expect(component).toMatchSnapshot();
});

test('correctly identifies the key clicked and passes that information to the callback', () => {
    const mockedTriggerAttackRelease = jest.fn()
    global.instrumentReferences = {
        '9132467532619432': {
            triggerAttackRelease: mockedTriggerAttackRelease
        }
    };
    const component = shallow(
        <PianoKeyLayer 
            pianoKeyLayerRef={React.createRef()}
            pitchesArray={createPitchesArray()}
            channelId="9132467532619432"
        />
    );
    const C5Key = component.find(Rect).find({ pitch: 'C5' });
    C5Key.simulate('click', {});
    expect(mockedTriggerAttackRelease).toHaveBeenCalledTimes(1);
    expect(mockedTriggerAttackRelease).toHaveBeenCalledWith('C5', '8n');
});