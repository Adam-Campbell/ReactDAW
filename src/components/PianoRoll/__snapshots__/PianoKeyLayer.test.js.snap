// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Layer
  y={40}
>
  <Rect
    fill="#e0e0e0"
    height={16}
    key="0"
    onClick={[Function]}
    pitch="B8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={0}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="1"
    onClick={[Function]}
    pitch="A#8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={16}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="2"
    onClick={[Function]}
    pitch="A8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={32}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="3"
    onClick={[Function]}
    pitch="G#8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={48}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="4"
    onClick={[Function]}
    pitch="G8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={64}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="5"
    onClick={[Function]}
    pitch="F#8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={80}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="6"
    onClick={[Function]}
    pitch="F8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={96}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="7"
    onClick={[Function]}
    pitch="E8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={112}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="8"
    onClick={[Function]}
    pitch="D#8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={128}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="9"
    onClick={[Function]}
    pitch="D8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={144}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="10"
    onClick={[Function]}
    pitch="C#8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={160}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="11"
    onClick={[Function]}
    pitch="C8"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={176}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="12"
    onClick={[Function]}
    pitch="B7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={192}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="13"
    onClick={[Function]}
    pitch="A#7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={208}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="14"
    onClick={[Function]}
    pitch="A7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={224}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="15"
    onClick={[Function]}
    pitch="G#7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={240}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="16"
    onClick={[Function]}
    pitch="G7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={256}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="17"
    onClick={[Function]}
    pitch="F#7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={272}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="18"
    onClick={[Function]}
    pitch="F7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={288}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="19"
    onClick={[Function]}
    pitch="E7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={304}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="20"
    onClick={[Function]}
    pitch="D#7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={320}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="21"
    onClick={[Function]}
    pitch="D7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={336}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="22"
    onClick={[Function]}
    pitch="C#7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={352}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="23"
    onClick={[Function]}
    pitch="C7"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={368}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="24"
    onClick={[Function]}
    pitch="B6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={384}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="25"
    onClick={[Function]}
    pitch="A#6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={400}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="26"
    onClick={[Function]}
    pitch="A6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={416}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="27"
    onClick={[Function]}
    pitch="G#6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={432}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="28"
    onClick={[Function]}
    pitch="G6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={448}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="29"
    onClick={[Function]}
    pitch="F#6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={464}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="30"
    onClick={[Function]}
    pitch="F6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={480}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="31"
    onClick={[Function]}
    pitch="E6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={496}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="32"
    onClick={[Function]}
    pitch="D#6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={512}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="33"
    onClick={[Function]}
    pitch="D6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={528}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="34"
    onClick={[Function]}
    pitch="C#6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={544}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="35"
    onClick={[Function]}
    pitch="C6"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={560}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="36"
    onClick={[Function]}
    pitch="B5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={576}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="37"
    onClick={[Function]}
    pitch="A#5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={592}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="38"
    onClick={[Function]}
    pitch="A5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={608}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="39"
    onClick={[Function]}
    pitch="G#5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={624}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="40"
    onClick={[Function]}
    pitch="G5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={640}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="41"
    onClick={[Function]}
    pitch="F#5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={656}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="42"
    onClick={[Function]}
    pitch="F5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={672}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="43"
    onClick={[Function]}
    pitch="E5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={688}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="44"
    onClick={[Function]}
    pitch="D#5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={704}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="45"
    onClick={[Function]}
    pitch="D5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={720}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="46"
    onClick={[Function]}
    pitch="C#5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={736}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="47"
    onClick={[Function]}
    pitch="C5"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={752}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="48"
    onClick={[Function]}
    pitch="B4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={768}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="49"
    onClick={[Function]}
    pitch="A#4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={784}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="50"
    onClick={[Function]}
    pitch="A4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={800}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="51"
    onClick={[Function]}
    pitch="G#4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={816}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="52"
    onClick={[Function]}
    pitch="G4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={832}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="53"
    onClick={[Function]}
    pitch="F#4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={848}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="54"
    onClick={[Function]}
    pitch="F4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={864}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="55"
    onClick={[Function]}
    pitch="E4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={880}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="56"
    onClick={[Function]}
    pitch="D#4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={896}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="57"
    onClick={[Function]}
    pitch="D4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={912}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="58"
    onClick={[Function]}
    pitch="C#4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={928}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="59"
    onClick={[Function]}
    pitch="C4"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={944}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="60"
    onClick={[Function]}
    pitch="B3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={960}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="61"
    onClick={[Function]}
    pitch="A#3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={976}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="62"
    onClick={[Function]}
    pitch="A3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={992}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="63"
    onClick={[Function]}
    pitch="G#3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1008}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="64"
    onClick={[Function]}
    pitch="G3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1024}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="65"
    onClick={[Function]}
    pitch="F#3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1040}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="66"
    onClick={[Function]}
    pitch="F3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1056}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="67"
    onClick={[Function]}
    pitch="E3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1072}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="68"
    onClick={[Function]}
    pitch="D#3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1088}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="69"
    onClick={[Function]}
    pitch="D3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1104}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="70"
    onClick={[Function]}
    pitch="C#3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1120}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="71"
    onClick={[Function]}
    pitch="C3"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1136}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="72"
    onClick={[Function]}
    pitch="B2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1152}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="73"
    onClick={[Function]}
    pitch="A#2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1168}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="74"
    onClick={[Function]}
    pitch="A2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1184}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="75"
    onClick={[Function]}
    pitch="G#2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1200}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="76"
    onClick={[Function]}
    pitch="G2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1216}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="77"
    onClick={[Function]}
    pitch="F#2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1232}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="78"
    onClick={[Function]}
    pitch="F2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1248}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="79"
    onClick={[Function]}
    pitch="E2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1264}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="80"
    onClick={[Function]}
    pitch="D#2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1280}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="81"
    onClick={[Function]}
    pitch="D2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1296}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="82"
    onClick={[Function]}
    pitch="C#2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1312}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="83"
    onClick={[Function]}
    pitch="C2"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1328}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="84"
    onClick={[Function]}
    pitch="B1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1344}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="85"
    onClick={[Function]}
    pitch="A#1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1360}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="86"
    onClick={[Function]}
    pitch="A1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1376}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="87"
    onClick={[Function]}
    pitch="G#1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1392}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="88"
    onClick={[Function]}
    pitch="G1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1408}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="89"
    onClick={[Function]}
    pitch="F#1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1424}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="90"
    onClick={[Function]}
    pitch="F1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1440}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="91"
    onClick={[Function]}
    pitch="E1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1456}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="92"
    onClick={[Function]}
    pitch="D#1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1472}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="93"
    onClick={[Function]}
    pitch="D1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1488}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="94"
    onClick={[Function]}
    pitch="C#1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1504}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="95"
    onClick={[Function]}
    pitch="C1"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1520}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="96"
    onClick={[Function]}
    pitch="B0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1536}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="97"
    onClick={[Function]}
    pitch="A#0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1552}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="98"
    onClick={[Function]}
    pitch="A0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1568}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="99"
    onClick={[Function]}
    pitch="G#0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1584}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="100"
    onClick={[Function]}
    pitch="G0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1600}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="101"
    onClick={[Function]}
    pitch="F#0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1616}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="102"
    onClick={[Function]}
    pitch="F0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1632}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="103"
    onClick={[Function]}
    pitch="E0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1648}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="104"
    onClick={[Function]}
    pitch="D#0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1664}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="105"
    onClick={[Function]}
    pitch="D0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1680}
  />
  <Rect
    fill="#47426c"
    height={16}
    key="106"
    onClick={[Function]}
    pitch="C#0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1696}
  />
  <Rect
    fill="#e0e0e0"
    height={16}
    key="107"
    onClick={[Function]}
    pitch="C0"
    stroke="#201826"
    strokeWidth={2}
    type="pianoKeyRect"
    width={48}
    x={0}
    y={1712}
  />
</Layer>
`;
